scenario/scripts/snapshot.py

```diff
@@ -0,0 +1,895 @@
+        )
+        meta_path = f"/var/lib/juju/agents/unit-{target.normalized}/charm/metadata.yaml"
+
+    raw_meta = _juju_ssh(
+        target,
+        f"cat {meta_path}",
+        model=model.name,
+    )
+    return yaml.safe_load(raw_meta)
+
+
+class RemotePebbleClient:
+    """Clever little class that wraps calls to a remote pebble client."""
+
+    # TODO: there is a .pebble.state
+    #  " j ssh --container traefik traefik/0 cat var/lib/pebble/default/.pebble.state | jq"
+    #  figure out what it's for.
+
+    def __init__(
+        self, container: str, target: JujuUnitName, model: Optional[str] = None
+    ):
+        self.socket_path = f"/charm/containers/{container}/pebble.socket"
+        self.container = container
+        self.target = target
+        self.model = model
+
+    def _run(self, cmd: str) -> str:
+        _model = f" -m {self.model}" if self.model else ""
+        command = f"juju ssh{_model} --container {self.container} {self.target.unit_name} /charm/bin/pebble {cmd}"
+        proc = run(shlex.split(command), capture_output=True)
+        if proc.returncode == 0:
+            return proc.stdout.decode("utf-8")
+        raise RuntimeError(
+            f"error wrapping pebble call with {command}: "
+            f"process exited with {proc.returncode}; "
+            f"stdout = {proc.stdout}; "
+            f"stderr = {proc.stderr}"
+        )
+
+    def can_connect(self) -> bool:
+        try:
+            version = self.get_system_info()
+        except Exception:
+            return False
+        return bool(version)
+
+    def get_system_info(self):
+        return self._run("version")
+
+    def get_plan(self) -> dict:
+        plan_raw = self._run("plan")
+        return yaml.safe_load(plan_raw)
+
+    def pull(
+        self, path: str, *, encoding: Optional[str] = "utf-8"
+    ) -> Union[BinaryIO, TextIO]:
+        raise NotImplementedError()
+
+    def list_files(
+        self, path: str, *, pattern: Optional[str] = None, itself: bool = False
+    ) -> List[ops.pebble.FileInfo]:
+        raise NotImplementedError()
+
+    def get_checks(
+        self,
+        level: Optional[ops.pebble.CheckLevel] = None,
+        names: Optional[Iterable[str]] = None,
+    ) -> List[ops.pebble.CheckInfo]:
+        _level = f" --level={level}" if level else ""
+        _names = (" " + f" ".join(names)) if names else ""
+        out = self._run(f"checks{_level}{_names}")
+        if out == "Plan has no health checks.":
+            return []
+        raise NotImplementedError()
+
+
+def fetch_file(
+    target: JujuUnitName,
+    remote_path: str,
+    container_name: str,
+    local_path: Path = None,
+    model: Optional[str] = None,
+) -> Optional[str]:
+    """Download a file from a live unit to a local path."""
+    # copied from jhack
+    model_arg = f" -m {model}" if model else ""
+    cmd = f"juju ssh --container {container_name}{model_arg} {target.unit_name} cat {remote_path}"
```

Comment:
can't recall the path to this solution, but it was long and painful. 
probably that scp isn't container-aware, so it's useless in this case.

Reply:
will add a note

Reply:
> can't recall the path to this solution, but it was long and painful. probably that scp isn't container-aware, so it's useless in this case.

What do mean by "scp isn't container-aware"? Does it not support pulling files from k8s charms, and only machine charms. Looking at the output of `juju help scp` for juju version 3.1.0-genericlinux-amd64, it seems to indicate that it is possible to pull files from containers using `juju scp ...`:

```text
# Copy a file ('chunks-inspect') from localhost to /loki directory
# in a specific container in a juju unit running in Kubernetes:
juju scp --container loki chunks-inspect loki-k8s/0:/loki
```

Maybe give it another try? If it still does not work, base64 encoding should be fine. I do something similar in cleantest, along though I encapsulate the target file/directory in a tarball to prevent it from being mangled by Python:

https://github.com/NucciTheBoss/cleantest/blob/29564ab114bda16bd1abaf2928c5450eb24b4e84/src/cleantest/data/file.py#L51-L91

^ I don't like how complicated I made this code (I am refactoring large parts of cleantest before releasing 0.4.0), so that is why I am pushing to see if `juju scp ...` works :sweat_smile: